### PR TITLE
tgui-next

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
This ports tgui-next from tg

transfer valves not done because we dont have them anywhere (now)

account database machine deleted because we dont use it

those two botany machines disabled because holy shit botany code bad
i've copied all the interfaces that have been nexted on tg and done a few of the nano->tgui-next and tgui->next conversions myself
our id computer is nothing like the tg one
without porting tg botany i'm not wasting my time converting those seed dna machine interfaces from nano
in terms of code ported from tg we have
smartfridge is now almost straight tg code
portgen is now almost straight tg code

## Why It's Good For The Game
better ui

## Changelog
:cl:
add: Added tgui-next
del: Removes nanoui
/:cl:

Closes #2726

image
image
image
image
image
image
image
image
image

### CODE REVIEW

**Comment by Qustinnus** [comment]
looks ok to me

File: code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -1152,6 +1152,13 @@ GLOBAL_LIST_INIT(wallitems, typecacheof(list(
 	pixel_x = initialpixelx
 	pixel_y = initialpixely
 
+/atom/proc/contains(atom/A)
+	if(!A)
+		return 0
+	for(var/atom/location = A.loc, location, location = location.loc)
+		if(location == src)
+			return 1
```
Comment (Rohesie) [inline code/__HELPERS/unsorted.dm]:
```suggestion
			return TRUE
```

File: code/game/objects/items/transfer_valve.dm [hunk 2]
```diff
@@ -40,7 +40,7 @@
 			to_chat(user, "<span class='notice'>You attach the tank to the transfer valve.</span>")
 
 		update_icon()
-		SSnano.update_uis(src) // update all UIs attached to src
+//		SSnano.update_uis(src) // update all UIs attached to src
```
Comment (Rohesie) [inline code/game/objects/items/transfer_valve.dm]: Delet

File: code/game/objects/machinery/computer/card.dm [hunk 3]
```diff
@@ -71,10 +71,10 @@
 		C.forceMove(src)
 		modify = C
 
-	SSnano.update_uis(src)
+	updateUsrDialog()
 	attack_hand(user)
 
-
+/*
```
Comment (Rohesie) [inline code/game/objects/machinery/computer/card.dm]: Delete

File: code/__HELPERS/unsorted.dm (hunk 1 above)
Comment (spookydonut) [reply code/__HELPERS/unsorted.dm]: ripped from tg

File: code/game/objects/items/tanks/tanks.dm [hunk 4]
```diff
@@ -70,6 +70,7 @@
 
 	ui_interact(user)
 
+/*
```
[the remark this answers is not in the document]
Comment (spookydonut) [reply code/game/objects/items/tanks/tanks.dm]: oh this is actually a todo, i need to convert this nanoui

File: code/game/objects/machinery/computer/card.dm (hunk 3 above)
Comment (spookydonut) [reply code/game/objects/machinery/computer/card.dm]: its a todo, i'll need this code if i'm ever going to convert the id computer

File: code/controllers/subsystem/tgui.dm [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+/datum/controller/subsystem/tgui/PreInit()
+	basehtml = file2text('tgui-next/packages/tgui/public/tgui-main.html')
+
+/datum/controller/subsystem/tgui/Shutdown()
+	close_all_uis()
+
+/datum/controller/subsystem/tgui/stat_entry()
+	..("P:[processing_uis.len]")
+
+/datum/controller/subsystem/tgui/fire(resumed = 0)
```
Comment (Rohesie) [inline code/controllers/subsystem/tgui.dm]:
```suggestion
/datum/controller/subsystem/tgui/fire(resumed = FALSE)
```

File: code/game/objects/items/portable_vendor.dm [hunk 6]
```diff
@@ -85,58 +91,48 @@
...
-		ui.set_initial_data(data)
-		ui.open()
-		ui.set_auto_update(1)
-
-
-/obj/item/portable_vendor/Topic(href, href_list)
-	. = ..()
-	if(.)
+/obj/item/portable_vendor/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/game/objects/items/portable_vendor.dm]:
```suggestion
	. = ..()
	if(.)
```

File: code/game/objects/items/tanks/tanks.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rohesie) [reply code/game/objects/items/tanks/tanks.dm]: Copy

File: code/game/objects/machinery/computer/card.dm (hunk 3 above)
Comment (Rohesie) [reply code/game/objects/machinery/computer/card.dm]: Alrighty.

File: code/game/objects/machinery/computer/crew.dm [hunk 7]
```diff
@@ -108,21 +98,26 @@
...
-	if(!ui)
-		ui = new(user, src, ui_key, "crew_monitor.tmpl", "Crew Monitoring Computer", 900, 800)
+	return data
 
-		// adding a template with the key "mapContent" enables the map ui functionality
-//		ui.add_template("mapContent", "crew_monitor_map_content.tmpl")
-		// adding a template with the key "mapHeader" replaces the map header content
-//		ui.add_template("mapHeader", "crew_monitor_map_header.tmpl")
+/obj/machinery/computer/crew/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/game/objects/machinery/computer/crew.dm]:
```suggestion
	. = ..()
	if(.)
```

File: code/game/objects/machinery/computer/dropship_weapons.dm [hunk 8]
```diff
@@ -12,7 +12,7 @@
 	var/obj/structure/dropship_equipment/selected_equipment //the currently selected equipment installed on the shuttle this console controls.
 	var/list/shuttle_equipments = list() //list of the equipments on the shuttle this console controls
 
-
+/*
```
Comment (Rohesie) [inline code/game/objects/machinery/computer/dropship_weapons.dm]: To do as well?

File: code/game/objects/machinery/self_destruct.dm [hunk 9]
```diff
@@ -40,12 +41,22 @@
...
+		ui = new(user, src, ui_key, "selfdestruct", name, ui_x, ui_y, master_ui, state)
+		ui.open()
+
+/obj/machinery/self_destruct/console/ui_data(mob/user)
+	return list("dest_status" = active_state)
 
-	switch(href_list["command"])
+
+/obj/machinery/self_destruct/console/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/game/objects/machinery/self_destruct.dm]:
```suggestion
	. = ..()
	if(.)
```

File: code/game/objects/machinery/sentries.dm [hunk 10]
```diff
@@ -364,34 +374,19 @@
...
-		ui.open()
-		ui.set_auto_update(1)
-
-/obj/machinery/marine_turret/Topic(href, href_list)
-	. = ..()
-	if(.)
-		return
-	if(usr.stat)
+/obj/machinery/marine_turret/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/game/objects/machinery/sentries.dm]:
```suggestion
	. = ..()
	if(.)
```

File: code/game/objects/machinery/vending/new_marine_vendors.dm [hunk 11]
```diff
@@ -118,116 +128,106 @@
 		"current_m_points" = m_points,
 		"displayed_records" = display_list,
 	)
+	return data
 
-	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+/obj/machinery/marine_selector/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/game/objects/machinery/vending/new_marine_vendors.dm]:
```suggestion
	. = ..()
	if(.)
```

File: code/game/objects/machinery/vending/vending.dm [hunk 12]
```diff
@@ -382,54 +388,48 @@
...
-		ui = new(user, src, ui_key, "vending_machine.tmpl", name , 450, 600)
-		ui.set_initial_data(data)
-		ui.open()
-		ui.set_auto_update(1)
-
-/obj/machinery/vending/Topic(href, href_list)
-	. = ..()
-	if(.)
+/obj/machinery/vending/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/game/objects/machinery/vending/vending.dm]:
```suggestion
	. = ..()
	if(.)
```

File: code/game/objects/objs.dm [hunk 13]
```diff
@@ -255,6 +255,21 @@
...
+	if(machine)
+		machine.on_unset_machine(src)
+		machine = null
+
+//called when the user unsets the machine.
+/atom/movable/proc/on_unset_machine(mob/user)
+	return
+
+/mob/proc/set_machine(obj/O)
+	if(src.machine)
```
Comment (Rohesie) [inline code/game/objects/objs.dm]:
```suggestion
	if(machine)
```

File: code/game/objects/objs.dm [hunk 14]
```diff
@@ -255,6 +255,21 @@
...
+		unset_machine()
+	src.machine = O
```
Comment (Rohesie) [inline code/game/objects/objs.dm]:
```suggestion
	machine = O
```

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 15]
```diff
@@ -342,60 +345,85 @@
...
+	data["isBeakerLoaded"] = beaker ? TRUE : FALSE
+	var/beakerContents = list()
+	if(beaker && beaker.reagents && beaker.reagents.reagent_list.len)
+		for(var/datum/reagent/R in beaker.reagents.reagent_list)
+			beakerContents += list(list("name" = R.name, "volume" = R.volume))
+	data["beakerContents"] = beakerContents
+	return data
+
+/obj/machinery/atmospherics/components/unary/cryo_cell/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]:
```suggestion
	. = ..()
	if(.)
```

File: code/modules/power/apc.dm [hunk 16]
```diff
@@ -699,20 +708,69 @@
...
+		return (val==1) ? 0 : val
+	else if(val == 3)
+		return 1
+	else
+		return 0
+
+/obj/machinery/power/apc/proc/can_use(mob/user, loud = 0) //used by attack_hand() and Topic()
+	if(IsAdminGhost(user))
+		return TRUE
+	if(isAI(user) && src.aidisabled)
```
Comment (Rohesie) [inline code/modules/power/apc.dm]:
```suggestion
	if(isAI(user) && aidisabled)
```

File: code/modules/power/apc.dm [hunk 17]
```diff
@@ -699,20 +708,69 @@
...
 
+/obj/machinery/power/apc/proc/setsubsystem(val)
+	if(cell && cell.charge > 0)
+		return (val==1) ? 0 : val
+	else if(val == 3)
+		return 1
+	else
+		return 0
+
+/obj/machinery/power/apc/proc/can_use(mob/user, loud = 0) //used by attack_hand() and Topic()
```
Comment (Rohesie) [inline code/modules/power/apc.dm]:
```suggestion
/obj/machinery/power/apc/proc/can_use(mob/user, loud = FALSE) //used by attack_hand() and Topic()
```

File: code/modules/power/apc.dm [hunk 18]
```diff
@@ -699,20 +708,69 @@
...
+	if(IsAdminGhost(user))
+		return TRUE
+	if(isAI(user) && src.aidisabled)
+		if(!loud)
+			to_chat(user, "<span class='danger'>\The [src] has eee disabled!</span>")
+		return FALSE
+	return TRUE
+
+/obj/machinery/power/apc/ui_act(action, params)
+	if(..() || !can_use(usr, 1) || locked)
```
Comment (Rohesie) [inline code/modules/power/apc.dm]:
```suggestion
	. = ..()
	if(. || !can_use(usr, 1) || locked)
```

File: code/modules/power/apc.dm [hunk 19]
```diff
@@ -699,20 +708,69 @@
...
+			else if(params["env"])
+				environ = setsubsystem(text2num(params["env"]))
+				update_icon()
+				update()
+			. = TRUE
+		if("overload")
+			if(usr.has_unlimited_silicon_privilege)
+				overload_lighting()
+				. = TRUE
+	return 1
```
Comment (Rohesie) [inline code/modules/power/apc.dm]: Why all these checks just to `return 1`?

File: code/modules/power/port_gen.dm [hunk 20]
```diff
@@ -147,117 +152,137 @@
...
+	name = "\improper M.R.S.P.A.C.M.A.N.-type portable generator"
+	base_icon = "portgen2"
+	icon_state = "portgen2_0"
+	circuit = /obj/item/circuitboard/machine/pacman/mrs
+	sheet_path = /obj/item/stack/sheet/mineral/diamond
+	power_gen = 40000
+	time_per_sheet = 80
+
+/obj/machinery/power/port_gen/pacman/mrs/overheat()
+	explosion(src.loc, 4, 4, 4, -1)
```
Comment (Rohesie) [inline code/modules/power/port_gen.dm]:
```suggestion
	explosion(loc, 4, 4, 4, -1)
```

File: code/modules/power/port_gen.dm [hunk 21]
```diff
@@ -147,117 +152,137 @@
...
-	power_gen = 40000
-	time_per_sheet = 120
-	board_path = /obj/item/circuitboard/machine/pacman/super
-	overheat()
-		explosion(src.loc, 3, 3, 3, -1)
+	power_gen = 15000
+	time_per_sheet = 85
+
+/obj/machinery/power/port_gen/pacman/super/overheat()
+	explosion(src.loc, 3, 3, 3, -1)
```
Comment (Rohesie) [inline code/modules/power/port_gen.dm]:
```suggestion
	explosion(loc, 3, 3, 3, -1)
```

File: code/modules/power/smes.dm [hunk 22]
```diff
@@ -250,37 +250,89 @@
...
-		ui.set_auto_update(1)
-
+		"outputLevel_text" = DisplayPower(output_level),
+		"outputLevelMax" = output_level_max,
+		"outputUsed" = DisplayPower(output_used)
+	)
+	return data
+
+/obj/machinery/power/smes/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/modules/power/smes.dm]:
```suggestion
	. =..()
	if(.)
```

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 23]
```diff
@@ -94,62 +116,79 @@
...
-		// open the new ui window
-		ui.open()
+	data["recordingRecipe"] = recording_recipe
+	return data
 
-/obj/machinery/chem_dispenser/Topic(href, href_list)
-	. = ..()
-	if(.)
+/obj/machinery/chem_dispenser/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/modules/reagents/Chemistry-Machinery.dm]:
```suggestion
	. =..()
	if(.)
```

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 24]
```diff
@@ -196,10 +252,9 @@
 /obj/machinery/chem_dispenser/beer
 	icon_state = "booze_dispenser"
 	name = "booze dispenser"
-	ui_title = "Booze Portal 9001"
-	energy = 100
-	accept_glass = 1
-	max_energy = 100
+	//ui_title = "Booze Portal 9001"
```
Comment (Rohesie) [inline code/modules/reagents/Chemistry-Machinery.dm]: This can be removed, no?

File: code/modules/shuttle/marine_dropship.dm [hunk 25]
```diff
@@ -423,14 +432,34 @@
...
+	return data
 
-	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
-	if (!ui)
-		ui = new(user, src, ui_key, "dropship_pilot_console.tmpl", "Pilot Control", 500, 600)
-		ui.set_initial_data(data)
-		ui.open()
-		ui.set_auto_update(1)
+/obj/machinery/computer/shuttle/marine_dropship/ui_act(action, params)
+	if(..())
```
Comment (Rohesie) [inline code/modules/shuttle/marine_dropship.dm]:
```suggestion
	. =..()
	if(.)
```

File: code/game/objects/items/portable_vendor.dm (hunk 6 above)
Comment (spookydonut) [reply code/game/objects/items/portable_vendor.dm]: No this is the style of tgui/tgui-next

File: code/game/objects/machinery/computer/dropship_weapons.dm (hunk 8 above)
Comment (spookydonut) [reply code/game/objects/machinery/computer/dropship_weapons.dm]: cough, totally

File: code/modules/power/apc.dm (hunk 19 above)
Comment (spookydonut) [reply code/modules/power/apc.dm]: ask tg

File: code/modules/reagents/Chemistry-Machinery.dm (hunk 24 above)
Comment (spookydonut) [reply code/modules/reagents/Chemistry-Machinery.dm]: I cant remember why i did this

**Comment by Rohesie** [comment]
Vendors seem broken. I've asked the people reporting it to comment here about it.

**Comment by Isy232** [comment]
None of the vendors seem to be working, they all load empty
image

File: code/game/objects/items/tanks/tanks.dm [hunk 26]
```diff
@@ -44,7 +44,7 @@
 			else
 				descriptive = "furiously hot"
 
-		to_chat(user, "<span class='notice'>\The [icon2html(src, user)][src] feels [descriptive]</span>")
+		to_chat(user, "<span class='notice'>\The [icon2html(src, user)][src] feels [descriptive], the guage reads [return_pressure()] kPa.</span>")
```
Comment (Rohesie) [inline code/game/objects/items/tanks/tanks.dm]:
```suggestion
		to_chat(user, "<span class='notice'>\The [icon2html(src, user)][src] feels [descriptive], the gauge reads [return_pressure()] kPa.</span>")
```

File: code/modules/admin/SDQL2/SDQL_2.dm [hunk 27]
```diff
@@ -1135,8 +1135,6 @@ GLOBAL_DATUM_INIT(sdql2_vv_statobj, /obj/effect/statclick/SDQL2_VV_all, new(null
 				v = SSmapping
 			if("SSmobs")
 				v = SSmobs
```
Comment (Rohesie) [inline code/modules/admin/SDQL2/SDQL_2.dm]:
```suggestion
				v = SSmobs
			if("SStgui"
				v = SStgui
```

File: code/modules/client/preferences_savefile.dm [hunk 28]
```diff
@@ -240,6 +245,8 @@
 	WRITE_FILE(S["focus_chat"], focus_chat)
 	WRITE_FILE(S["clientfps"], clientfps)
 	WRITE_FILE(S["tooltips"], tooltips)
+	WRITE_FILE(S["tgui_fancy"], tgui_fancy)
+	WRITE_FILE(S["tgui_lock"], tgui_lock)
```
Comment (Rohesie) [inline code/modules/client/preferences_savefile.dm]: File size needs to be changed.

File: code/modules/power/apc.dm (hunk 18 above)
Comment (Rohesie) [reply code/modules/power/apc.dm]: Edited the suggestion to preserve what's still valid.

File: code/modules/power/apc.dm (hunk 19 above)
Comment (Rohesie) [reply code/modules/power/apc.dm]: Please don't just resolve things like this.
The return seems to determine whether the UI should update or not, so this is safe to remove.

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 29]
```diff
@@ -94,62 +116,79 @@
...
 		data["beakerMaxVolume"] = beaker.volume
+		data["beakerTransferAmounts"] = beaker.possible_transfer_amounts
 	else
 		data["beakerCurrentVolume"] = null
 		data["beakerMaxVolume"] = null
+		data["beakerTransferAmounts"] = null
 
-	var chemicals[0]
-	for (var/re in dispensable_reagents)
+	var/chemicals[0]
```
Comment (Rohesie) [inline code/modules/reagents/Chemistry-Machinery.dm]:
```suggestion
	var/list/chemicals = list()
```

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 30]
```diff
@@ -174,12 +213,29 @@
 	icon_state = "soda_dispenser"
 	name = "soda fountain"
 	desc = "A drink fabricating machine, capable of producing many sugary drinks with just one touch."
-	ui_title = "Soda Dispens-o-matic"
-	energy = 100
-	accept_glass = TRUE
+	//ui_title = "Soda Dispens-o-matic"
```
Comment (Rohesie) [inline code/modules/reagents/Chemistry-Machinery.dm]: Remove?

**Comment by Spyroshark** [comment]
image

Seems like there's something weird with vendors that take coins, and that any item with quotation marks doesn't have it's name show right

**Comment by Rohesie** [comment]
There's also #3169

**Comment by XSlayer300** [comment]
Crew monitoring computer is not working. http://prntscr.com/pqvrb2

**Comment by Mark9013100** [comment]
I only get a black screen when trying to look at the Crew Monitor's/SMES' UI, kept getting this post clearing cache.

**Comment by Rohesie** [comment]
Fun > Launch Pod Supply seems to be doing nothing.

**Comment by SandPoot** [comment]
Bug report:
get to a automated weapons closet
open it
leave, do not leave the 1 tile distance though
if someone gets something from it, the window will re-open for you.

**Comment by Isy232** [comment]
image
Something is up with windows

**Comment by Rohesie** [comment]
Sorry, @Isy232 , that one was for #3200

**Comment by Metroidlover** [comment]
image
SuperPacman Generators are busted

**Pull request merged**